File: prioris/core/axes.py
"""Normalized scoring axes and planning attributes.

Pure module: no I/O, no external dependency.
"""
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum


# ---------------------------------------------------------------- axes score
class Axis(str, Enum):
    BLK = "BLK"  # real blockage           0..5
    CDR = "CDR"  # delay cost              0..4
    IMP = "IMP"  # impact                  0..4
    IRR = "IRR"  # irreversibility         0..3
    INA = "INA"  # one-month inaction      0..4
    HOR = "HOR"  # visibility horizon      0..4
    ALN = "ALN"  # goal alignment          0..3


AXIS_MAX: dict[Axis, int] = {
    Axis.BLK: 5, Axis.CDR: 4, Axis.IMP: 4, Axis.IRR: 3,
    Axis.INA: 4, Axis.HOR: 4, Axis.ALN: 3,
}

# Value used by the uncertainty dampener.
# Intentionally conservative for ALN/IRR to avoid inflating importance.
AXIS_MEDIAN: dict[Axis, int] = {
    Axis.BLK: 2, Axis.CDR: 2, Axis.IMP: 2, Axis.IRR: 1,
    Axis.INA: 2, Axis.HOR: 2, Axis.ALN: 1,
}

# Each label should be a natural answer to its axis question. It is both a UI
# button and the scale provided to the LLM for NLU.
AXIS_LABELS: dict[Axis, list[str]] = {
    Axis.BLK: ["Personne", "Moi seul", "Une autre personne",
               "Une équipe ou plusieurs personnes", "Un acteur critique",
               "Plusieurs équipes ou une chaîne critique"],
    Axis.CDR: ["Rien — le coût ne bouge pas", "Il s'accumule doucement",
               "Il s'accumule nettement", "Il s'aggrave de plus en plus",
               "Falaise : tout se joue à une date"],
    Axis.IMP: ["Négligeable", "Un peu de confort en plus",
               "Une différence notable", "Une différence majeure",
               "Structurant pour la suite"],
    Axis.IRR: ["Réversible à tout moment", "Rattrapable avec effort",
               "Rattrapable jusqu'à une date", "Irréversible"],
    Axis.INA: ["Rien du tout", "Une gêne", "Un vrai problème",
               "Une crise", "Des dégâts irrécupérables"],
    Axis.HOR: ["Jamais", "Dans plus d'un mois", "Dans 2 à 4 semaines",
               "Cette semaine", "C'est déjà visible"],
    Axis.ALN: ["Aucun objectif", "Contribution indirecte",
               "Contribution directe", "Contribution majeure"],
}

AXIS_QUESTIONS: dict[Axis, str] = {
    Axis.INA: "Si personne n'y touche pendant un mois, que se passe-t-il concrètement ?",
    Axis.BLK: "Qui est bloqué si ce n'est pas fait cette semaine ?",
    Axis.IMP: "Quelle différence réelle entre « fait » et « pas fait » ?",
    Axis.HOR: "Quand le problème deviendra-t-il visible ?",
    Axis.CDR: "Comment le coût évolue-t-il si tu attends ?",
    Axis.IRR: "Peut-on revenir en arrière ou rattraper plus tard ?",
    Axis.ALN: "Cette tâche contribue-t-elle à un de tes objectifs de vie ?",
}


class Incertitude(int, Enum):
    SUR = 0
    HESITANT = 1
    NE_SAIT_PAS = 2


# --------------------------------------------- planning attributes
class Estimation(str, Enum):
    LT15 = "<15 min"
    M15_30 = "15–30 min"
    M30_60 = "30–60 min"
    H1_2 = "1–2 h"
    H2_4 = "2–4 h"
    GT4 = ">4 h"
    INCONNUE = "inconnue"


# Median minutes. INCONNUE maps to 60, but the evaluation is provisional and
# the task is never scheduled in a plan.
ESTIMATION_MIN: dict[Estimation, int] = {
    Estimation.LT15: 10, Estimation.M15_30: 22, Estimation.M30_60: 45,
    Estimation.H1_2: 90, Estimation.H2_4: 180, Estimation.GT4: 300,
    Estimation.INCONNUE: 60,
}


class Effort(int, Enum):
    FAIBLE = 1
    MOYEN = 2
    ELEVE = 3


class Demandeur(str, Enum):
    MOI = "moi"
    COLLEGUE = "collegue"
    MANAGER = "manager"
    CLIENT = "client"


class Priorite(str, Enum):
    P1 = "P1"
    P2 = "P2"
    P3 = "P3"
    P4 = "P4"

    @property
    def niveau(self) -> int:
        return int(self.value[1])


# Quadrant legend used by renderers and buttons.
QUADRANT_INFO: dict[str, dict] = {
    "Q1": {"p": "P1", "emoji": "🔥", "nom": "Urgent et important",
           "action": "faire en premier"},
    "Q2": {"p": "P2", "emoji": "🎯", "nom": "Important, pas urgent",
           "action": "planifier — le quadrant des objectifs"},
    "Q3": {"p": "P3", "emoji": "⚡", "nom": "Urgent, pas important",
           "action": "déléguer ou traiter vite et petit"},
    "Q4": {"p": "P4", "emoji": "🗑", "nom": "Ni urgent ni important",
           "action": "reporter ou abandonner sans culpabilité"},
}

PRIORITE_LABELS: dict[Priorite, str] = {
    Priorite(info["p"]): f"{info['p']} {info['emoji']} {info['nom']}"
    for info in QUADRANT_INFO.values()
}


@dataclass(frozen=True)
class Metadata:
    """Metadata outside the score, used only by bias detection.

    In express interviews, these keep neutral defaults; dependent biases are
    simply not evaluated.
    """
    demandeur: Demandeur = Demandeur.MOI
    visibilite: int = 0          # 0..3
    pression: int = 0            # 0..3
    temps_investi_h: float = 0.0
    subjective: Priorite | None = None


# ----------------------------------------------- express defaults
def hor_from_deadline(deadline_days: int | None) -> int:
    """Derive HOR from the deadline in express mode. No deadline means median."""
    if deadline_days is None:
        return AXIS_MEDIAN[Axis.HOR]
    if deadline_days <= 0:
        return 4
    if deadline_days <= 7:
        return 3
    if deadline_days <= 30:
        return 2
    return 1


def express_defaults(deadline_days: int | None) -> dict[Axis, int]:
    """Deterministic defaults for axes not asked in express mode.

    IMP is deliberately absent: impact and inaction are independent concepts,
    so every interview asks IMP explicitly.
    """
    return {
        Axis.HOR: hor_from_deadline(deadline_days),
        Axis.IRR: 1,
    }

File: rust/src/core/axes.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum Axis {
    BLK,
    CDR,
    IMP,
    IRR,
    INA,
    HOR,
    ALN,
}

impl Axis {
    pub const ALL: [Self; 7] = [
        Self::BLK,
        Self::CDR,
        Self::IMP,
        Self::IRR,
        Self::INA,
        Self::HOR,
        Self::ALN,
    ];

    pub const fn code(self) -> &'static str {
        match self {
            Self::BLK => "BLK",
            Self::CDR => "CDR",
            Self::IMP => "IMP",
            Self::IRR => "IRR",
            Self::INA => "INA",
            Self::HOR => "HOR",
            Self::ALN => "ALN",
        }
    }

    pub const fn max(self) -> u8 {
        match self {
            Self::BLK => 5,
            Self::CDR | Self::IMP | Self::INA | Self::HOR => 4,
            Self::IRR | Self::ALN => 3,
        }
    }

    pub const fn median(self) -> u8 {
        match self {
            Self::BLK | Self::CDR | Self::IMP | Self::INA | Self::HOR => 2,
            Self::IRR | Self::ALN => 1,
        }
    }

    pub const fn question_fr(self) -> &'static str {
        match self {
            Self::INA => "Si personne n'y touche pendant un mois, que se passe-t-il concrètement ?",
            Self::BLK => "Qui est bloqué si ce n'est pas fait cette semaine ?",
            Self::IMP => "Quelle différence réelle entre fait et pas fait ?",
            Self::HOR => "Quand le problème deviendra-t-il visible ?",
            Self::CDR => "Comment le coût évolue-t-il si tu attends ?",
            Self::IRR => "Peut-on revenir en arrière ou rattraper plus tard ?",
            Self::ALN => "Cette tâche contribue-t-elle à un de tes objectifs de vie ?",
        }
    }

    pub const fn labels_fr(self) -> &'static [&'static str] {
        match self {
            Self::BLK => &[
                "Personne",
                "Moi seul",
                "Une autre personne",
                "Une équipe ou plusieurs personnes",
                "Un acteur critique",
                "Plusieurs équipes ou une chaîne critique",
            ],
            Self::CDR => &[
                "Rien",
                "Il s'accumule doucement",
                "Il s'accumule nettement",
                "Il s'aggrave",
                "Falaise à une date",
            ],
            Self::IMP => &[
                "Négligeable",
                "Un peu de confort",
                "Différence notable",
                "Différence majeure",
                "Structurant",
            ],
            Self::IRR => &[
                "Réversible",
                "Rattrapable avec effort",
                "Rattrapable jusqu'à une date",
                "Irréversible",
            ],
            Self::INA => &[
                "Rien",
                "Une gêne",
                "Un vrai problème",
                "Une crise",
                "Dégâts irrécupérables",
            ],
            Self::HOR => &[
                "Jamais",
                "Dans plus d'un mois",
                "Dans 2 à 4 semaines",
                "Cette semaine",
                "Déjà visible",
            ],
            Self::ALN => &[
                "Aucun objectif",
                "Contribution indirecte",
                "Contribution directe",
                "Contribution majeure",
            ],
        }
    }
}

impl std::str::FromStr for Axis {
    type Err = String;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        Self::ALL
            .into_iter()
            .find(|axis| axis.code().eq_ignore_ascii_case(value))
            .ok_or_else(|| format!("unknown axis: {value}"))
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[repr(u8)]
pub enum Uncertainty {
    #[default]
    Certain = 0,
    Hesitant = 1,
    Unknown = 2,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
pub enum Estimate {
    Lt15,
    M15_30,
    M30_60,
    H1_2,
    H2_4,
    Gt4,
    #[default]
    Unknown,
}

impl Estimate {
    pub const fn minutes(self) -> u32 {
        match self {
            Self::Lt15 => 10,
            Self::M15_30 => 22,
            Self::M30_60 => 45,
            Self::H1_2 => 90,
            Self::H2_4 => 180,
            Self::Gt4 => 300,
            Self::Unknown => 60,
        }
    }

    pub const fn db_value(self) -> &'static str {
        match self {
            Self::Lt15 => "<15 min",
            Self::M15_30 => "15–30 min",
            Self::M30_60 => "30–60 min",
            Self::H1_2 => "1–2 h",
            Self::H2_4 => "2–4 h",
            Self::Gt4 => ">4 h",
            Self::Unknown => "inconnue",
        }
    }
}

impl std::str::FromStr for Estimate {
    type Err = String;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        match value {
            "<15 min" => Ok(Self::Lt15),
            "15–30 min" | "15-30 min" => Ok(Self::M15_30),
            "30–60 min" | "30-60 min" => Ok(Self::M30_60),
            "1–2 h" | "1-2 h" => Ok(Self::H1_2),
            "2–4 h" | "2-4 h" => Ok(Self::H2_4),
            ">4 h" => Ok(Self::Gt4),
            "inconnue" | "unknown" | "" => Ok(Self::Unknown),
            _ => Err(format!("unknown estimate: {value}")),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[repr(u8)]
pub enum Effort {
    Low = 1,
    #[default]
    Medium = 2,
    High = 3,
}

impl Effort {
    pub fn from_u8(value: u8) -> Self {
        match value {
            1 => Self::Low,
            3 => Self::High,
            _ => Self::Medium,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Priority {
    P1,
    P2,
    P3,
    P4,
}

impl Priority {
    pub const fn level(self) -> i8 {
        match self {
            Self::P1 => 1,
            Self::P2 => 2,
            Self::P3 => 3,
            Self::P4 => 4,
        }
    }

    pub const fn as_str(self) -> &'static str {
        match self {
            Self::P1 => "P1",
            Self::P2 => "P2",
            Self::P3 => "P3",
            Self::P4 => "P4",
        }
    }
}

impl std::str::FromStr for Priority {
    type Err = String;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        match value.to_ascii_uppercase().as_str() {
            "P1" => Ok(Self::P1),
            "P2" => Ok(Self::P2),
            "P3" => Ok(Self::P3),
            "P4" => Ok(Self::P4),
            _ => Err(format!("unknown priority: {value}")),
        }
    }
}

pub fn horizon_from_deadline(deadline_days: Option<i64>) -> u8 {
    match deadline_days {
        None => Axis::HOR.median(),
        Some(days) if days <= 0 => 4,
        Some(days) if days <= 7 => 3,
        Some(days) if days <= 30 => 2,
        Some(_) => 1,
    }
}
